File: Problems/boj_괄호.py
#https://www.acmicpc.net/problem/9012
#Data Structure

import sys
input = sys.stdin.readline
t = int(input())

for _ in range(t):
    vps = input().rstrip()
        
    stack = []
    for i in vps:
        if i=='(':
            stack.append(i)
        else:
            if stack:
                stack.pop()
            else:
                print('NO')
                break
    else:
        if not stack:
            print('YES')
        else:
            print('NO')

File: Problems/cpp/boj_괄호.cpp
//https://www.acmicpc.net/problem/9012
//Data Structure

#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;

    for (int i=0; i<t; i++){
        string input;
        cin >> input;

        bool flag = false;
        vector<char> stk;

        for (const auto& c : input){
            if (c=='('){
                stk.push_back(c);
            }
            else{
                if (!stk.empty()){
                    stk.pop_back();
                }
                else{
                    cout << "NO" << '\n';
                    flag = true;
                    break;
                }
            }
        }
        
        if (flag) continue;
        
        if(stk.empty())
            cout << "YES" << '\n';
        else
            cout << "NO" << '\n';
    }
}
